src/test/java/net/snowflake/client/jdbc/cloud/storage/GcmEncryptionProviderTest.java

```diff
@@ -0,0 +1,169 @@
+import java.nio.file.Files;
+import java.security.InvalidAlgorithmParameterException;
+import java.security.InvalidKeyException;
+import java.security.NoSuchAlgorithmException;
+import java.security.SecureRandom;
+import java.util.Base64;
+import javax.crypto.BadPaddingException;
+import javax.crypto.IllegalBlockSizeException;
+import javax.crypto.NoSuchPaddingException;
+import net.snowflake.client.jdbc.MatDesc;
+import net.snowflake.common.core.RemoteStoreFileEncryptionMaterial;
+import org.apache.commons.io.FileUtils;
+import org.apache.commons.io.IOUtils;
+import org.junit.Before;
+import org.junit.Test;
+import org.mockito.ArgumentCaptor;
+
+public class GcmEncryptionProviderTest {
+  private final SecureRandom random = new SecureRandom();
+
+  private final ArgumentCaptor<StorageObjectMetadata> storageObjectMetadataArgumentCaptor =
+      ArgumentCaptor.forClass(StorageObjectMetadata.class);
+  private final ArgumentCaptor<MatDesc> matDescArgumentCaptor =
+      ArgumentCaptor.forClass(MatDesc.class);
+  private final ArgumentCaptor<byte[]> cekIvDataArgumentCaptor =
+      ArgumentCaptor.forClass(byte[].class);
+  private final ArgumentCaptor<byte[]> kekIvDataArgumentCaptor =
+      ArgumentCaptor.forClass(byte[].class);
+  private final ArgumentCaptor<byte[]> encKekArgumentCaptor = ArgumentCaptor.forClass(byte[].class);
+  private final ArgumentCaptor<byte[]> keyAadArgumentCaptor = ArgumentCaptor.forClass(byte[].class);
+  private final ArgumentCaptor<byte[]> dataAadArgumentCaptor =
+      ArgumentCaptor.forClass(byte[].class);
+  private final ArgumentCaptor<Long> contentLengthArgumentCaptor =
+      ArgumentCaptor.forClass(Long.class);
+
+  private final StorageObjectMetadata meta = mock(StorageObjectMetadata.class);
+  private final SnowflakeStorageClient storageClient = mock(SnowflakeStorageClient.class);
+
+  private final String queryStageMasterKey =
+      Base64.getEncoder().encodeToString(random.generateSeed(16));
+  private final RemoteStoreFileEncryptionMaterial encMat = new RemoteStoreFileEncryptionMaterial();
+
+  byte[] plainText = "the quick brown fox jumps over the lazy dog".getBytes(StandardCharsets.UTF_8);
+  byte[] dataAad = "data aad".getBytes(StandardCharsets.UTF_8);
+  byte[] keyAad = "key aad".getBytes(StandardCharsets.UTF_8);
+
+  @Before
+  public void setUp() {
+    encMat.setQueryStageMasterKey(queryStageMasterKey);
+    encMat.setSmkId(123);
+    encMat.setQueryId("query-id");
+  }
+
+  @Test
+  public void testEncryptAndDecryptStreamWithoutAad() throws Exception {
+    InputStream plainTextStream = new ByteArrayInputStream(plainText);
+
+    byte[] cipherText = encryptStream(plainTextStream, null, null);
+
+    InputStream inputStream = decryptStream(cipherText, null, null);
+    byte[] decryptedPlainText = IOUtils.toByteArray(inputStream);
+    assertArrayEquals(plainText, decryptedPlainText);
+  }
+
+  @Test
+  public void testEncryptAndDecryptStreamWithAad() throws Exception {
```

Comment:
Would also be interesting to test cases with large AAD, eg. 1GB

Reply:
AAD can sometimes be confusing and/or forgotten about. Here I'd like to see a test where the AAD exceeds the length allowed by the underlying library. That is, testing "max AAD length" + 1 in order to see what the failure looks like. 

So, this is much more about how it fails.

Reply:
What if we restrict the AAD size to something like 2^16?

Reply:
What do you mean by restricting? This is not a customer value, this is a value that we set in driver. We have full control over it and now it's going to be empty. So we may have something like that:
```
byte[] aad = new byte[0];
if (aad.length > 65k) {
   // throw some exception
}
```
But I believe this is not something you want?

Reply:
I know that this library is going to be used by a limited audience at first, but often it can grow over time without the context we're discussing now. Given that the AAD is required for decryption, it's important that we have a strong opinion about how it might be used and enforce that. eg., I think we should:

 - Have an opinion on the max size for the AAD.
 - Enforce that max size.

Otherwise, there may be future cases where someone does use a large AAD and it leads to surprises. I don't have a specific concern today, but this is a strategy for futureproofing this implemntation.